### PR TITLE
Draft a deployment notifier that posts on PR pre-prod status

### PR DESCRIPTION
**Pull request by jhchabran**
Fixes https://github.com/sourcegraph/sourcegraph/issues/31839 

This introduces a deployment notifier, that will detect the changes being deployed, post GitHub comments and post a Slack summary. 

As we're one step from turning this into something usable right now to notify production deployments, I went ahead and added specific code able to handle production deployments. 

1. It takes as an input a commit coming from the CI pipeline that triggers the preprod pipeline.2. 
    - A `sourcegraph.guess-commit` flag is provided in order to analyze the changes to detect which commit is going to be deployed, so we can use this on production deployment which are driven by Renovate. 
3. It gets the list of commits that separates that commit from the one currently deployed on the staging. 
4. It posts on each associated PR a messages.
    - Skipped for now. 
5. It posts a Slack message with a list of PR and pings their author. 

See also https://github.com/sourcegraph/deploy-sourcegraph-cloud/pull/15859 

@bobheadxi I think a safe way to start with this is to just use with Slack notifications for a few days, then to wire the GitHub comments if we see it working correctly. With the fact that some deployments are done sometime once in a day in the case of `gitserver`, that would mean a lot of PR being commented on, so better be safe than sorry for posting 200 comments 😆 

A remaining minor question is how to ship the binary on the other `sourcegraph/deploy-*` repo, but that shouldn't be a problem. 

## Test plan

- [x] Added unit tests with recorded payloads
- [x] Run with `-pretend` on `deploy-sourcegraph-cloud` in a branch 
- Run with real Slack and GitHub comments on small subset of commits.

### CODE REVIEW

**Comment by jhchabran** [comment]
@bobheadxi we could eventually use this start giving partial updates on deployments, by using this tool inside the production pipeline, if and only if we're running the daily deployment job, so we're sure that the entirety of the apps have been deployed. 

Or, because we're using comments, we could list which images got updated and therefore provide an accurate picture to our users of their deployment status, right now.

**Comment by bobheadxi** [comment]
Did you mean to link this PR to https://github.com/sourcegraph/sourcegraph/issues/31838 instead?

> Or, because we're using comments, we could list which images got updated and therefore provide an accurate picture to our users of their deployment status, right now.

This sounds like the best approach! The author can then assess if the service their change affects has been rolled out. An additional comment can happen later when more services roll out the change, or we could update the original comment

**Comment by jhchabran** [comment]
Test run, mocking the live version commit to avoid running through more than 100 commits to gather PRs https://buildkite.com/sourcegraph/deploy-sourcegraph-cloud/builds/196257#9a59bf6d-fc72-4d77-81c7-9c31841fcea1/70-94

File: enterprise/dev/deployment-notifications/main.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	SlackAnnounceWebhook   string
+}
+
+func (f *Flags) Parse() {
+	flag.StringVar(&f.GitHubToken, "github.token", os.Getenv("GITHUB_TOKEN"), "mandatory github token")
+	flag.StringVar(&f.SourcegraphCommit, "sourcegraph.commit", "", "Sourcegraph commit being deployed")
+	flag.StringVar(&f.Environment, "environment", "", "Environment being deployed")
+	flag.StringVar(&f.MockLiveCommit, "mock.live-commit", "", "Use this commit instead of requesting the commit deployed on the target environment")
+	flag.BoolVar(&f.Pretend, "pretend", false, "Pretend to post notifications, printing to stdout instead")
+	flag.BoolVar(&f.GuessSourcegraphCommit, "sourcegraph.guess-commit", false, "Attempt at deducting the deployed commit from the changes in the diff")
```
Comment (jhchabran) [inline enterprise/dev/deployment-notifications/main.go:37]: TODO s/deducting/deducing|infering/

File: enterprise/dev/deployment-notifications/deployments_notifier.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	repoName            = "sourcegraph"
+	commitsPerPage      = 30
+	maxCommitsPageCount = 5
+)
+
+var (
+	ErrAlreadyDeployed = errors.New("target commit matches live commit")
+)
+
+type DeploymentNotifier struct {
```
Comment (jhchabran) [inline enterprise/dev/deployment-notifications/deployments_notifier.go:26]: I'm not sure to like this name, up for suggestions here :)

File: enterprise/dev/deployment-notifications/README.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+## How it works
+
+Deployment notifier works as following:
+
+1. Request the commit currently running on the target environment
+2. Compute the deployed applications based on the changes included in the latest commit.
+3. Find all PRs that happened in between the commit found in 1. and the commit being currently deployed (`-sourcegraph.commit`)
+4. Post a comment in each of those PRs with the list of applications computed in 2.
+5. Post a Slack message pinging all pull request authors.
```
Comment (taylorsperry) [inline enterprise/dev/deployment-notifications/README.md:22]: This is probably a dumb question, but if I'm running the deployment notifier on one of my PRs, what are "all PRs that happened in between the commit [in the target environment]" and my last commit on my PR? What other PRs would I have in my branch that aren't already in the target environment?

Comment (jhchabran) [reply enterprise/dev/deployment-notifications/README.md:22]: Let's say we have the following:

- commit 10 running in production
- commits 11, 12, 13 are present on the `main` branch at t=8:00AM 
- commit 14 is committed on the `main` branch at t=9:00AM 

The way deployments are working is as following: 

- on a recurring basis, Renovate will look if there are new applications available to deploy, by looking at the images that are the result of running the CI on the `main` branch. 
- it will update the `deploy-sourcegraph-cloud` repository with those new image as a result. 

Let's say that Renovate is running at t=8:55AM. 

Deployment Notifier is going to see that commit 13 is the one being deployed. It's also going to see that the commit in production is commit 10. It means that it not only needs to notify about commit 13 but also commits 11 and 12. Each commit on `main` is the result of a single PR being merged. So the PRs associated to 11 and 12 are _all PRs that happened in between_. 

It's possible that by the time the CI for a `deploy-sourcegraph-cloud` run, we have another commit, commit 14 to be present on the `main` branch, because it took some time to get started and we're now at t=9:01AM. In that case, we don't want commit 14, it would be incorrect. This is why the deployment notifier is referring to a target commit, commit 13 in this example, to know where to stop.

Comment (taylorsperry) [reply enterprise/dev/deployment-notifications/README.md:22]: Perfect, thank you @jhchabran. I think the Renovate part was the piece I was missing 👍 🧠

File: enterprise/dev/deployment-notifications/guess_commit.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+)
+
+var imageCommitRegexp = `(?m)^\+\s+image:\s[^/]+\/sourcegraph\/APPNAME:\d{6}_\d{4}-\d{2}-\d{2}_([^@]+)@sha256.*$` // (?m) stands for multiline.
+
+func guessSourcegraphCommit() (string, error) {
+	files := []string{
+		// If we're looking at a continuous deployment, we'll always find the frontend being updated.
+		"base/frontend/sourcegraph-frontend-internal.Deployment.yaml",
+		// Else, it's either a daily deployment with giterver or Tomas bumping it.
+		"base/gitserver/gitserver.StatefulSet.yaml",
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifications/guess_commit.go:17]: Doesn't repo-updater roll out daily as well?

File: enterprise/dev/deployment-notifications/main.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+	SlackAnnounceWebhook   string
+}
+
+func (f *Flags) Parse() {
+	flag.StringVar(&f.GitHubToken, "github.token", os.Getenv("GITHUB_TOKEN"), "mandatory github token")
+	flag.StringVar(&f.SourcegraphCommit, "sourcegraph.commit", "", "Sourcegraph commit being deployed")
+	flag.StringVar(&f.Environment, "environment", "", "Environment being deployed")
+	flag.StringVar(&f.MockLiveCommit, "mock.live-commit", "", "Use this commit instead of requesting the commit deployed on the target environment")
+	flag.BoolVar(&f.Pretend, "pretend", false, "Pretend to post notifications, printing to stdout instead")
+	flag.BoolVar(&f.GuessSourcegraphCommit, "sourcegraph.guess-commit", false, "Attempt at inferring the deployed commit from the changes in the diff")
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifications/main.go:37]:
```suggestion
	flag.BoolVar(&f.InferSourcegraphCommit, "sourcegraph.infer-commit", false, "Attempt at inferring the deployed commit from the changes in the diff")
```

You already use `infer` here and I think it's the much better term!

File: enterprise/dev/deployment-notifications/slack.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+	err = tmpl.Execute(&sb, presenter)
+	if err != nil {
+		return "", err
+	}
+
+	return sb.String(), nil
+}
+
+// postSlackUpdate attempts to send the given summary to at each of the provided webhooks.
+func postSlackUpdate(webhook string, summary string) error {
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifications/slack.go:78]: We have so many copies of this we need a library 😆 not part of this PR though

File: enterprise/dev/deployment-notifications/guess_commit.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+				continue
+			}
+		} else {
+			return "", err
+		}
+	}
+	return "", nil
+}
+
+func extractCommitFromDiff(output []byte, appname string) string {
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifications/guess_commit.go:35]:
```suggestion
func extractSourcegraphCommitFromDeploymentManifestsDiff(output []byte, appname string) string {
```

File: enterprise/dev/deployment-notifications/README.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+```sh
+deployment-notifier -environment $MY_ENV -sourcegraph.commit $MY_COMMIT
+```
+
+### Flags 
+
+- `-github.token` (defaults to `$GITHUB_TOKEN`) 
+- `-sourcegraph.commit` (optional) the SHA1 of the commit being deployed, used to find the pull requests to mention.
+- `-sourcegraph.guess-commit` (optional) infers the SHA1 of the commit being deployed from the changes, supersedes `-sourcegraph.commit`.
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifications/README.md:18]: When will this be used?

**Review by bobheadxi — COMMENTED** [review]
First pass review! Havent looked too closely at the notification part yet

**Comment by bobheadxi** [comment]
@jhchabran I think this fixes https://github.com/sourcegraph/sourcegraph/issues/32154, _not_ https://github.com/sourcegraph/sourcegraph/issues/31839? The two issues seem very similar

File: enterprise/dev/deployment-notifications/README.md (hunk 8 above)
Comment (jhchabran) [reply enterprise/dev/deployment-notifications/README.md:18]: It's meant to be used with the production deployments, basically, anything that depends on Renovate.

File: enterprise/dev/deployment-notifications/main.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+	SlackToken             string
+	SlackAnnounceWebhook   string
+}
+
+func (f *Flags) Parse() {
+	flag.StringVar(&f.GitHubToken, "github.token", os.Getenv("GITHUB_TOKEN"), "mandatory github token")
+	flag.StringVar(&f.SourcegraphCommit, "sourcegraph.commit", "", "Sourcegraph commit being deployed")
+	flag.StringVar(&f.Environment, "environment", "", "Environment being deployed")
+	flag.StringVar(&f.MockLiveCommit, "mock.live-commit", "", "Use this commit instead of requesting the commit deployed on the target environment")
+	flag.BoolVar(&f.Pretend, "pretend", false, "Pretend to post notifications, printing to stdout instead")
```
[the remark this answers is not in the document]
Comment (jhchabran) [reply enterprise/dev/deployment-notifications/main.go:36]: I think `--pretend` comes from my Gentoo days, a loooong time ago 😆

File: enterprise/dev/deployment-notifications/guess_commit.go (hunk 4 above)
Comment (jhchabran) [reply enterprise/dev/deployment-notifications/guess_commit.go:17]: Ah right, the comment is slightly confusing here. It's just one of the many manifests being deployed daily. We just need a reliable way to guess the commit being applied, so I had to chose one.

**Comment by jhchabran** [comment]
> @jhchabran I think this fixes #32154, _not_ #31839? The two issues seem very similar

Yeah, the former is a sub-issue of the later. We'll close both :)

File: enterprise/dev/deployment-notifications/slack.go (hunk 6 above)
Comment (jhchabran) [reply enterprise/dev/deployment-notifications/slack.go:78]: Heavy +1 :D

File: enterprise/dev/deployment-notifications/README.md (hunk 8 above)
Comment (bobheadxi) [reply enterprise/dev/deployment-notifications/README.md:18]: This should just be the default behaviour, and `-sourcegraph.commit` can disable the the inference

File: enterprise/dev/deployment-notifications/deployments_notifier.go (hunk 2 above)
Comment (bobheadxi) [reply enterprise/dev/deployment-notifications/deployments_notifier.go:26]: If the name is clunky, just `Notifier` should be sufficient

File: enterprise/dev/deployment-notifications/deployments_notifier.go [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+func (dn *DeploymentNotifier) getNewPullRequests(ctx context.Context, liveCommit string) ([]*github.PullRequest, error) {
+	repoCommits, err := dn.getNewCommits(ctx, liveCommit)
+	if err != nil {
+		return nil, err
+	}
+	var pullsSinceLastCommit []*github.PullRequest
+	for _, rc := range repoCommits {
+		pulls, _, err := dn.ghc.PullRequests.ListPullRequestsWithCommit(
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifications/deployments_notifier.go:82]: I wonder if we can save an API call here by simply parsing the PR number from the commit message

File: enterprise/dev/deployment-notifications/deployments_notifier.go [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+		targetCommit: targetCommit,
+	}
+}
+
+// getNewCommits returns a slice of commits starting from the target commit up to the currently deployed commit.
+func (dn *DeploymentNotifier) getNewCommits(ctx context.Context, lastCommit string) ([]*github.RepositoryCommit, error) {
+	var page = 1
+	var commits []*github.RepositoryCommit
+	for page != 0 && page != maxCommitsPageCount {
+		cs, resp, err := dn.ghc.Repositories.ListCommits(ctx, repoOwner, repoName, &github.CommitsListOptions{
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifications/deployments_notifier.go:50]:
```suggestion
		cs, resp, err := dn.ghc.Repositories.ListCommits(ctx, repoOwner, sourceRepoName, &github.CommitsListOptions{
```

etc.

Comment (bobheadxi) [reply enterprise/dev/deployment-notifications/deployments_notifier.go:50]: It's awkward because the tool is in `sourcegraph/sourcegraph`, but runs in `deploy-sourcegraph-cloud, but also refers back the the `sourcegraph` repo - we should be very explicit wherever possible

File: enterprise/dev/deployment-notifications/deployments_notifier.go [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+	return &DeploymentNotifier{
+		vr:           vr,
+		ghc:          ghc,
+		changedFiles: changedFiles,
+		targetCommit: targetCommit,
+	}
+}
+
+// getNewCommits returns a slice of commits starting from the target commit up to the currently deployed commit.
+func (dn *DeploymentNotifier) getNewCommits(ctx context.Context, lastCommit string) ([]*github.RepositoryCommit, error) {
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifications/deployments_notifier.go:46]:
```suggestion
func (dn *DeploymentNotifier) getNewSourceCommits(ctx context.Context, sinceCommit string) ([]*github.RepositoryCommit, error) {
```

File: enterprise/dev/deployment-notifications/deployments_notifier.go [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+			if commit.GetSHA() == lastCommit {
+				return commits[currentCommitIdx:i], nil
+			}
+		}
+		page = resp.NextPage
+	}
+	return nil, errors.Newf("commit %s not found in the last %d commits", lastCommit, maxCommitsPageCount*commitsPerPage)
+}
+
+func (dn *DeploymentNotifier) getNewPullRequests(ctx context.Context, liveCommit string) ([]*github.PullRequest, error) {
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifications/deployments_notifier.go:75]:
```suggestion
func (dn *DeploymentNotifier) getNewSourcePullRequests(ctx context.Context, sinceCommit string) ([]*github.PullRequest, error) {
```

File: enterprise/dev/deployment-notifications/deployments_notifier.go [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+		return nil, err
+	}
+
+	if liveCommit == dn.targetCommit {
+		return nil, errors.Wrap(ErrAlreadyDeployed, dn.vr.Environment())
+	}
+
+	prs, err := dn.getNewPullRequests(ctx, liveCommit)
+	if err != nil {
+		return nil, err
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifications/deployments_notifier.go:111]:
```suggestion
		return nil, errors.Wrap(err, "get pull requests")
```

File: enterprise/dev/deployment-notifications/deployments_notifier.go [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+		}
+		pullsSinceLastCommit = append(pullsSinceLastCommit, pulls...)
+	}
+	return pullsSinceLastCommit, nil
+}
+
+func (dn *DeploymentNotifier) Report(ctx context.Context) (*report, error) {
+	liveCommit, err := dn.vr.LastCommit()
+	if err != nil {
+		return nil, err
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifications/deployments_notifier.go:102]:
```suggestion
		return nil, errors.Wrap(err, "last commit")
```

File: enterprise/dev/deployment-notifications/deployments_notifier.go [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+		PullRequests:      prs,
+		DeployedAt:        time.Now().In(time.UTC).Format(time.RFC822Z),
+		Apps:              dn.deployedApps(),
+		BuildkiteBuildURL: os.Getenv("BUILDKITE_BUILD_URL"),
+	}
+
+	return &report, nil
+}
+
+func (dn *DeploymentNotifier) deployedApps() []string {
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifications/deployments_notifier.go:125]:
```suggestion
func (dn *DeploymentNotifier) deployedServices() []string {
```

We refer to these as `services` everywhere else - I've never seen anyone refer to frontend or gitserver as an "app"

File: enterprise/dev/deployment-notifications/infer_commit.go [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+	// There are multiple files being updated when a deployment PR occurs. We're picking those two here
+	// for no particular reason other than the chances are pretty low for these to change soon.
+	files := []string{
+		// If we're looking at a continuous deployment, we'll always find the frontend being updated.
+		"base/frontend/sourcegraph-frontend-internal.Deployment.yaml",
+		// Else, it's either a daily deployment with giterver or Tomas bumping it.
+		"base/gitserver/gitserver.StatefulSet.yaml",
+	}
+	for _, file := range files {
+		diffCommand := []string{"diff", "@^", file}
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifications/infer_commit.go:22]: Is this correct? If run on a manifest PR, this will just diff the most recent commit, which is not accurate

File: enterprise/dev/deployment-notifications/version_request.go [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+	"net/http"
+	"strings"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+type VersionRequester interface {
+	LastCommit() (string, error)
+	Environment() string
+}
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifications/version_request.go:14]: We have services that roll out at different cadences, should this interface reflect that? Right now it seems we assume a universal version to get deployed

Comment (bobheadxi) [reply enterprise/dev/deployment-notifications/version_request.go:14]: Judging from `inferSourcegraphCommit` we expect multi-service rollouts to work by luck. I think it's a better idea to ditch `__version` altogether, and rely entirely on the commits embedded in the image tags within the manifests

Comment (bobheadxi) [reply enterprise/dev/deployment-notifications/version_request.go:14]: This interface can then return `map[Service]Commit`, and we can perform far more accurate diffs based on our infrastructure-as-code.

File: enterprise/dev/deployment-notifications/infer_commit.go (hunk 17 above)
Comment (bobheadxi) [reply enterprise/dev/deployment-notifications/infer_commit.go:22]: Though I guess we might be just running this post-merge, though it's not immediately clear to me

File: enterprise/dev/deployment-notifications/infer_commit.go [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+func inferSourcegraphCommit() (string, error) {
+	// There are multiple files being updated when a deployment PR occurs. We're picking those two here
+	// for no particular reason other than the chances are pretty low for these to change soon.
+	files := []string{
+		// If we're looking at a continuous deployment, we'll always find the frontend being updated.
+		"base/frontend/sourcegraph-frontend-internal.Deployment.yaml",
+		// Else, it's either a daily deployment with giterver or Tomas bumping it.
+		"base/gitserver/gitserver.StatefulSet.yaml",
+	}
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifications/infer_commit.go:20]: With the appraoch in https://github.com/sourcegraph/sourcegraph/pull/32459#discussion_r831558825 , we should just use `images.DeploySourcegraphImages` to regex search for exactly what we are looking for

Comment (bobheadxi) [reply enterprise/dev/deployment-notifications/infer_commit.go:20]: IMO that comment + using the image lists would be both a clearer and far more deterministic detector for what services have changed in a given manifest diff, and a lot simpler in that there is less magic knowledge of "tendencies" within our changes

File: enterprise/dev/deployment-notifications/version_request.go (hunk 18 above)
Comment (bobheadxi) [reply enterprise/dev/deployment-notifications/version_request.go:14]: As noted in https://github.com/sourcegraph/sourcegraph/pull/32459#discussion_r831567916 this might not be that complicated thanks to our existing image lists, and it will also allow us to support cases like someone manually bumping a single service better

File: enterprise/dev/deployment-notifications/deployments_notifier.go [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+
+	"github.com/google/go-github/v41/github"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+var (
+	repoOwner           = "sourcegraph"
+	repoName            = "sourcegraph"
+	commitsPerPage      = 30
+	maxCommitsPageCount = 5
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply enterprise/dev/deployment-notifications/deployments_notifier.go:19]: Potential mitigation: https://github.com/sourcegraph/sourcegraph/pull/32459#discussion_r831545983

File: enterprise/dev/deployment-notifications/version_request.go (hunk 18 above)
Comment (bobheadxi) [reply enterprise/dev/deployment-notifications/version_request.go:14]: I left a similar comment in the Miro board, we should:

> leave comments on PRs whenever a commit deploys, with a list of _which_ services have deployed, and let the author figure out if it is relevant

File: enterprise/dev/deployment-notifications/README.md (hunk 8 above)
Comment (jhchabran) [reply enterprise/dev/deployment-notifications/README.md:18]: I think it's safer to have the script complain about missing info that can be inferred that having it magically run without you noticing how it's finding its stuff.

File: enterprise/dev/deployment-notifications/infer_commit.go (hunk 17 above)
Comment (jhchabran) [reply enterprise/dev/deployment-notifications/infer_commit.go:22]: Yes, I should document that, it's meant to be run post merge.

File: enterprise/dev/deployment-notifications/version_request.go (hunk 18 above)
Comment (jhchabran) [reply enterprise/dev/deployment-notifications/version_request.go:14]: It's already detecting which services are being deployed there, by looking at the changed files. I think you are confusing "how do we detect what's the new commit" with "how do we know which new services got deployed" here.

File: enterprise/dev/deployment-notifications/README.md (hunk 8 above)
Comment (bobheadxi) [reply enterprise/dev/deployment-notifications/README.md:18]: It makes the intention of the tool quite confusing unfortunately. I noted this in another comment, but I think we should remove the `__version` implementation completely because that one is misleading, given specific services can be deployed at varying cadences for different reasons

File: enterprise/dev/deployment-notifications/version_request.go (hunk 18 above)
Comment (bobheadxi) [reply enterprise/dev/deployment-notifications/version_request.go:14]: It makes the semantics of the interface confusing, and wrong sometimes - if we detect gitserver rolling out, then using the live version requester makes this wrong. We should just roll everything into one, and remove the guesswork in favour of explicitness. The architecture could go:

1. Run on postmerge over manifests
2. In the diff, look for each image in `DeploySourcegraphImages`, find references to that image and their corresponding commit based on their tag (which has commit info embedded), and populate `map[Service]{Previous, Next}`.
   a. At this point we have an accurate state to work with
4. Do the git log over the detected service diffs

File: enterprise/dev/deployment-notifications/deployments_notifier.go (hunk 11 above)
Comment (jhchabran) [reply enterprise/dev/deployment-notifications/deployments_notifier.go:104]: I think that with the last commits, it's way way less confusing now.

File: enterprise/dev/deployment-notifier/slack.go [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+
+	"github.com/sourcegraph/sourcegraph/dev/team"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+var slackTemplate = `:truck: *{{.Environment}}* deployment (<{{.BuildURL}}|build>)
+
+- Applications:
+{{- range .Services }}
+    - _{{ . }}_
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifier/slack.go:22]: Can we use code format with backticks here for service names? You might have to `fmt.Sprintf` it in in order to get around the template literal, or you might be able to just escape it somehow

File: enterprise/dev/deployment-notifier/deployments_notifier.go [hunk 22]
```diff
@@ -0,0 +1,208 @@
...
+	repoName            = "sourcegraph"
+	commitsPerPage      = 30
+	maxCommitsPageCount = 5
+)
+
+var (
+	ErrNoRelevantChanges = errors.New("no services changed, nothing to notify")
+)
+
+type DeploymentNotifier struct {
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifier/deployments_notifier.go:28]: nit: file name `deployments_notifier`, but struct is `DeploymentNotifier`. I think the singular noun here is more correct!

File: enterprise/dev/deployment-notifier/deployments_notifier.go [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+			if strings.HasPrefix(commit.GetSHA(), oldCommit) {
+				return commits[currentCommitIdx:i], nil
+			}
+		}
+		page = resp.NextPage
+	}
+	return nil, errors.Newf("commit %s not found in the last %d commits", oldCommit, maxCommitsPageCount*commitsPerPage)
+}
+
+func parsePRNumberInMergeCommit(message string) int {
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifier/deployments_notifier.go:129]: 😍

File: enterprise/dev/deployment-notifier/deployments_notifier_test.go [hunk 24]
```diff
@@ -0,0 +1,184 @@
...
+
+func TestParsePRNumberInMergeCommit(t *testing.T) {
+	tests := []struct {
+		name    string
+		message string
+		want    int
+	}{
+		{name: "Merge commit with revert", message: `Revert "Support diffing for unrelated commits. (#32015)" (#32737)`, want: 32737},
+		{name: "Normal commit", message: `YOLO I commit on main without PR`, want: 0},
+		{name: "Merge commit", message: `batches: Properly quote name in YAML (#32951)`, want: 32951},
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifier/deployments_notifier_test.go:175]: Stretch goal: commit messages with truncated PR numbers, e.g. 

- https://github.com/sourcegraph/sourcegraph/commit/666b1298ea3883bf3c6db209fe2f044cfedd5cde
- https://github.com/sourcegraph/sourcegraph/commit/a2d6a94781148f2f5a47e3bb81bba47af916a3cb

I think we just need to be resilient against trailing `...` - a simple handling could just be to discard _any_ truncated commit message

A more complex handling could be to "rejoin" the message and the body if there is `...` 😆

File: enterprise/dev/deployment-notifier/manifest_differ.go [hunk 25]
```diff
@@ -0,0 +1,153 @@
...
+		// If matching an addition, the + needs to be escaped to not be parsed as a
+		// count operator.
+		escapedOp = "\\+"
+	} else {
+		escapedOp = "-"
+	}
+
+	re := strings.ReplaceAll(imageCommitRegexp, "DIFF_OP", escapedOp)
+	return regexp.MustCompile(re)
+}
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifier/manifest_differ.go:102]: LGTM, though fingers crossed we don't run into any weird diffs 🙏 I think we can deal with that when/if it happens though, and worst case we get a botched notification or two, so this should be robust enough

File: enterprise/dev/deployment-notifier/README.md [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+- `-environment` either `preprod` or `production`
+- `-dry` (optional) do not post on Slack or GitHub, just print out what would be posted.
+- `-slack.token` Slack Token used to find the matching Slack handle for pull request authors.
+- `-slack.webhook` Slack webhook URL to post the notifications on.
+
+## How it works
+
+Deployment notifier works as following:
+
+1. Inspect the diff of current commit (supposed to be a merge commit)
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifier/README.md:26]:
```suggestion
1. Inspect the diff the deployment manifests' current commit (supposed to be a merge commit, e.g. run on a release or deploy branch)
```

File: enterprise/dev/deployment-notifier/README.md [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+2. Compute the deployed applications and take note of the old commit and new commit for each of them.
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifier/README.md:27]:
```suggestion
2. Compute the deployed applications and take note of the old and new Sourcegraph commit for each of them.
```

File: enterprise/dev/deployment-notifier/README.md [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+3. For each set of old and new commits, find all PRs that happened after the old commit, up to the new commit.
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifier/README.md:28]:
```suggestion
3. For each set of old and new Sourcegraph commits, find all PRs that happened after the old commit, up to the new commit.
```

File: enterprise/dev/deployment-notifier/README.md [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+4. Post a comment in each of those PRs with the list of applications computed in 2.
+5. Post a Slack message pinging all pull request authors.
+
+## Contributing
+
+### Local usage
+
+To avoid spamming comments, Deployment Notifier comes with a few flags to help:
+
+- `-dry` prints a summary of what would be posted on the standard output.
```
Comment (bobheadxi) [inline enterprise/dev/deployment-notifier/README.md:38]: 👨‍🍳

**Review by bobheadxi — APPROVED** [review]
This is fantastic - thank you for responding to all the feedback, and I think this new implementation is a lot cleaner 😁

I do think you accidentally committed the `deployment-notifications` binary though, might want to remove that!

File: enterprise/dev/deployment-notifier/slack.go (hunk 21 above)
Comment (jhchabran) [reply enterprise/dev/deployment-notifier/slack.go:22]: Yeah the escape is a bit annoying but it works, it's 

```go
mystr := `foobar` + "`" + `baz`
```

File: enterprise/dev/deployment-notifier/deployments_notifier_test.go (hunk 24 above)
Comment (jhchabran) [reply enterprise/dev/deployment-notifier/deployments_notifier_test.go:175]: The messages are only truncated in the UI, so we're good with this 👍
